### PR TITLE
Export visible trees as nml in frontend

### LINKED ISSUE #2094: Investigate NML import/export in frontend
To be able to export individual trees and merge tracings via drag'n'drop we need a frontend NML parser and generator. Let's see whether this is feasible.

See https://discuss.webknossos.org/t/save-extract-selected-trees/550

<a href="https://timer.scm.io/repos/537237ce1a00001a003c4a14/issues/2094/create?referer=github" target="_blank">Log Time</a>

### PR DESCRIPTION
**Pull request by daniel-wer**
Missing:
- [x] spinner while exporting
- [x] correct name for the nml 

### Mailable description of changes:
 - [TracingView] It is now possible to download an nml file of the currently visible trees. Select `Export as nml` from the dropdown in the trees tab.

### Steps to test:
- Create a tracing with multiple trees, a user bounding box, comments and branchpoints. Select `Export as nml` from the dropdown in the trees tab and save the nml. Select `Download` from the dropdown next to the save button and save the nml. Compare the two nmls -> they should be equal.
- Hide some of the trees via their checkmark and export as nml again. The hidden trees should not be part of the nml.
- Open a very big tracing, e.g. 200.000 nodes, the export functionality should still work reasonably fast.
- The nml name should be the same for the download and export functionality.

### Issues:
- fixes #2094

------
- [ ] Ready for review

### CODE REVIEW

**Comment by daniel-wer** [comment]
Currently the backend serializes integers like `1` as `1.0` (sometimes), whereas the frontend serializes `1` as `1`, I suppose that's not a problem.
A test would probably be cool to compare the nml the backend generates with the one the frontend generates. I'm not sure how to compare the two nmls properly as the order of nodes can be changed and so on...

**Comment by daniel-wer** [comment]
If someone can explain to me what we could use [this](https://github.com/scalableminds/nml-spec/blob/master/nml-spec.xsd) file here for, I'll maybe update it, but I don't see the benefit right now.

**Comment by normanrz** [comment]
We could use it for validation of user-uploaded NMLs (which @youri-k's PR already does) or perhaps to bootstrap an NML parser. Also it might be helpful for documentation purposes. Anyways, not high priority.

File: app/assets/javascripts/oxalis/model/helpers/nml_helpers.js [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+  ];
+}
+
+function serializeTag(
+  name: string,
+  properties: { [string]: ?(string | number | boolean) },
+  closed: boolean = true,
+): string {
+  return `<${name} ${Object.keys(properties)
+    .map(key => `${key}="${properties[key] != null ? properties[key].toString() : ""}"`)
```
Comment (hotzenklotz) [inline app/assets/javascripts/oxalis/model/helpers/nml_helpers.js]: Object to string. Perhaps`JSON.stringify`? BTW, stringify accepts "indentation".

File: app/assets/javascripts/oxalis/view/right-menu/trees_tab_view.js [hunk 2]
```diff
@@ -69,6 +79,27 @@ class TreesTabView extends React.PureComponent<Props> {
     Store.dispatch(toggleInactiveTreesAction());
   }
 
+  exportAsNml = async () => {
+    this.setState({ isExporting: true });
+    // Allow the spinner to render
+    await Utils.sleep(1);
+    const state = Store.getState();
+    const nml = serializeToNml(state, this.props.skeletonTracing);
+    const anchor = document.createElement("a");
```
[the remark this answers is not in the document]
Comment (normanrz) [reply app/assets/javascripts/oxalis/view/right-menu/trees_tab_view.js]: Perhaps we could just use https://www.npmjs.com/package/file-saver here

File: app/assets/javascripts/oxalis/model/helpers/nml_helpers.js [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+  const shortAnnotationId = tracing.annotationId.slice(-6);
+
+  return `${datasetName}__${tracingType}__${userName}__${shortAnnotationId}.nml`;
+}
+
+export function serializeToNml(state: OxalisState, tracing: SkeletonTracingType): string {
+  // Only visible trees will be serialized!
+  // _.filter throws flow errors here, because the type definitions are wrong and I'm not able to fix them
+  const visibleTrees = Object.keys(tracing.trees)
+    .filter(treeId => tracing.trees[+treeId].isVisible)
```
[the remark this answers is not in the document]
Comment (normanrz) [reply app/assets/javascripts/oxalis/model/helpers/nml_helpers.js]: I guess this is for flow? The more readable way of converting to number would be `Number(treeId)` imo

File: app/assets/javascripts/oxalis/model/helpers/nml_helpers.js [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+  OxalisState,
+  SkeletonTracingType,
+  NodeMapType,
+  EdgeType,
+  TreeType,
+} from "oxalis/store";
+
+function indent(array: Array<string>): Array<string> {
+  array.forEach((line, index) => {
+    array[index] = `  ${line}`;
```
[the remark this answers is not in the document]
Comment (daniel-wer) [reply app/assets/javascripts/oxalis/model/helpers/nml_helpers.js]: In my first iteration this was indeed a map, which I then changed to a for loop for a very decent performance increase. I'll add a comment explaining this.
`String.padStart` is not really fitting here, as I don't want to pad to a fixed length, but simply want to prepend two spaces.

File: app/assets/javascripts/oxalis/model/helpers/nml_helpers.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (daniel-wer) [reply app/assets/javascripts/oxalis/model/helpers/nml_helpers.js]: Yes as I indeed am following your reasoning @hotzenklotz , Flow doesn't ^^
I'll change it to `Number(...)`.

File: app/assets/javascripts/oxalis/model/helpers/nml_helpers.js (hunk 1 above)
Comment (daniel-wer) [reply app/assets/javascripts/oxalis/model/helpers/nml_helpers.js]: Not sure I understand the benefit here. `JSON.stringify` would yield sth. like `{"name":"test","x":3}`, whereas I am building sth. like `<tag name="test" x="3">`.

Comment (hotzenklotz) [reply app/assets/javascripts/oxalis/model/helpers/nml_helpers.js]: Right, that's probably not useful here.

**Comment by daniel-wer** [comment]
I should've addressed all of your feedback and made the necessary changes. The `file-saver` module works very well. I cannot tick the check boxes, but they should all be resolved except for the second one, which we'll fix if it turns out to be a problem I'd say.
Only thing I'm not happy about is the `await Utils.sleep(1000)` to wait for the progress-modal to render when triggering the nml download. I don't know how to do this more elegantly.

**Pull request merged**